vcd/resource_vcd_vapp_vm_tools.go

```diff
@@ -0,0 +1,1168 @@
+			}
+
+			if joinDomainUser, isSetJoinDomainUser := d.GetOkExists("customization.0.join_domain_user"); isSetJoinDomainUser {
+				customizationSection.DomainUserName = joinDomainUser.(string)
+			}
+
+			if joinDomainPasswd, isSetJoinDomainPasswd := d.GetOkExists("customization.0.join_domain_password"); isSetJoinDomainPasswd {
+				customizationSection.DomainUserPassword = joinDomainPasswd.(string)
+			}
+
+			if joinDomainOu, isSetJoinDomainOu := d.GetOkExists("customization.0.join_domain_account_ou"); isSetJoinDomainOu {
+				customizationSection.MachineObjectOU = joinDomainOu.(string)
+			}
+
+		}
+	}
+}
+
+// setGuestCustomizationData is responsible for persisting all guest customization details into statefile
+func setGuestCustomizationData(d *schema.ResourceData, vm *govcd.VM) error {
+	customizationSection, err := vm.GetGuestCustomizationSection()
+	if err != nil {
+		return fmt.Errorf("unable to get guest customization section: %s", err)
+	}
+
+	dSet(d, "computer_name", customizationSection.ComputerName)
+
+	customizationBlock := make([]interface{}, 1)
+	customizationBlockAttributes := make(map[string]interface{})
+
+	customizationBlockAttributes["enabled"] = customizationSection.Enabled
+	customizationBlockAttributes["change_sid"] = customizationSection.ChangeSid
+	customizationBlockAttributes["allow_local_admin_password"] = customizationSection.AdminPasswordEnabled
+	customizationBlockAttributes["must_change_password_on_first_login"] = customizationSection.ResetPasswordRequired
+	customizationBlockAttributes["auto_generate_password"] = customizationSection.AdminPasswordAuto
+	customizationBlockAttributes["admin_password"] = customizationSection.AdminPassword
+	customizationBlockAttributes["number_of_auto_logons"] = customizationSection.AdminAutoLogonCount
+	customizationBlockAttributes["join_domain"] = customizationSection.JoinDomainEnabled
+	customizationBlockAttributes["join_org_domain"] = customizationSection.UseOrgSettings
+	customizationBlockAttributes["join_domain_name"] = customizationSection.DomainName
+	customizationBlockAttributes["join_domain_user"] = customizationSection.DomainUserName
+	customizationBlockAttributes["join_domain_password"] = customizationSection.DomainUserPassword
+	customizationBlockAttributes["join_domain_account_ou"] = customizationSection.MachineObjectOU
+	customizationBlockAttributes["initscript"] = customizationSection.CustomizationScript
+
+	customizationBlock[0] = customizationBlockAttributes
+
+	err = d.Set("customization", customizationBlock)
+	if err != nil {
+		return fmt.Errorf("")
+	}
+
+	return nil
+}
+
+// handleExposeHardwareVirtualization toggles hardware virtualization according `expose_hardware_virtualization` field value.
+func handleExposeHardwareVirtualization(d *schema.ResourceData, newVm *govcd.VM) error {
+	// The below operation assumes VM is powered off and does not check for it because VM is being
+	// powered on in the last stage of create/update cycle
```

Comment:
```suggestion
	// The operation below assumes the VM is powered off and does not check for status because the VM is being
	// powered on in the last stage of create/update cycle
```

Reply:
Fixed.